### PR TITLE
Implement a very lazy translation of headers

### PR DESCRIPTION
**Pull request by pushkarnk**
This pull request proposes a very lazy translation of `HTTPHeaders` to `HeadersContainer`. The difference between these structures, that matters most, is the conformance to `Collection`. `HeadersContainer` is a `Collection`, `HTTPHeaders` isn't. 

Conforming `HTTPHeaders` to `Collection` via an extension isn't straightforward. A conformance to `Collection` using the existing API leads to a very inelegant and sub-optimal implementation.

In that case, the only alternative we are left with is an "as lazy as possible" translation. Inside `HeadersContainer`, we maintain an extra backing store of type `HTTPHeaders`. The existing backing store, a `Dictionary` lies dormant. Only when the user code tries to access the `HeadersContainer` as  `Collection`, a translation is done and the `Dictionary` backing store comes to life. Subsequently, we update both the backing stores. This is to make sure there's no need for the `HeadersContainer` to `HTTPHeaders` translation.

In short, the `HTTPHeaders` -> `HeadersContainer` is done very lazily and the vice versa isn't needed.

### CODE REVIEW

**Comment by pushkarnk** [comment]
On `TechEmpower/JSON` this yields a benefit of around 8-9%.

**Comment by pushkarnk** [comment]
Also, I've realised that though the `Kitura` tests have enough test coverage for `HeadersContainer`, the `Kitura-NIO` repo is negligible coverage. I'm planning to write tests as a part of this pull request.

**Comment by pushkarnk** [comment]
Looks like a change to `BlueSSLService` is breaking Kitura-NIO on Darwin

https://github.com/IBM-Swift/BlueSSLService/commit/468d581297ed6167eb9c7244cc707cc6ebc48be5#diff-92ba0ed683849bc916d3ff460ce0b4b6

**Comment by nethraravindran** [comment]
`BlueSSLService` bug has been fixed in https://github.com/IBM-Swift/BlueSSLService/pull/67 and tagged

**Comment by pushkarnk** [comment]
A summary of the perf gains this PR gives on the different TechEmpower benchmarks (column `lazy` indicates the new numbers). All measurements are in MB/s.


Plaintext | 1.0.7 | lazy | KituraNet
-- | -- | -- | --
256 | 12.02 | 14.5 | 14.83
1024 | 11.93 | 14.17 | 13.66
4096 | 11.56 | 14.16 | 12.21
16384 | 11.57 | 13.73 | 11.37
  |   |   |  
  |   |   |  
JSON | 1.0.7 | lazy | KituraNet
8 | 5.43 | 6.67 | 8.81
16 | 7.79 | 8.89 | 11.29
32 | 9.26 | 10.73 | 11.39
64 | 9.39 | 11.4 | 11.37
128 | 9.42 | 11.6 | 11.41
256 | 9.49 | 11.66 | 11.4
  |   |   |  
DB | 1.0.7 | lazy | KituraNet
8 | 2.96 | 3.18 | 3.59
16 | 4.39 | 4.79 | 5.21
32 | 4.82 | 5.24 | 6.04
64 | 4.87 | 5.41 | 6.61
128 | 4.9 | 5.56 | 6.81
256 | 5.07 | 5.64 | 6.82
  |   |   |  
Updates | 1.0.7 | lazy | KituraNet
1 | 2.96 | 3.1 | 5.16
5 | 1.71 | 1.71 | 1.8
10 | 1.39 | 1.4 | 1.43
15 | 1.27 | 1.27 | 1.29
25 | 1.16 | 1.16 | 1.18
  |   |   |  
Queries | 1.07 | lazy | KituraNet
1 | 3.94 | 4.18 | 5.12
5 | 2.63 | 2.71 | 3.02
10 | 2.41 | 2.45 | 2.58
15 | 2.29 | 2.31 | 2.4
20 | 2.24 | 2.23 | 2.3
25 | 2.2 | 2.17 | 2.25
